drivers: adc: :adc_ad559x: add various fixes and improvements

LukaszMadejGrinn opened the pull request:
Add the following fixes and improvements:

1. Add configuration of internal refference voltage value so raw ADC readings can be converted to mV utilizing dc_raw_to_millivolts() ADC API function.

2. Fix driver start with CONFIG_THREAD_NAME not enabled.

3. Improve ADC read function to make it robust.

GeorgeCGV commented on drivers/adc/adc_ad5592.c:
```diff
@@ -222,9 +244,11 @@ static int adc_ad5592_init(const struct device *dev)
-	ret = k_thread_name_set(tid, "adc_ad5592");
-	if (ret < 0) {
-		return ret;
+	if (IS_ENABLED(CONFIG_THREAD_NAME)) {
+		ret = k_thread_name_set(tid, "adc_ad5592");
+		if (ret < 0) {
+			return ret;
+		}
```
Wouldn't it be better to ignore the return value of `k_thread_name_set`? That way it will work fine when thread naming is off but tracing is on; and when both are on or off.

LukaszMadejGrinn replied:
Result can be ignored and it is done this way in number of files. After analysis of the `k_thread_name_set()` function one can say that the return value is useles (and it is in fact at the moment). On the other hand this is how the API looks like and my approach here was to be aligned to the API well. If dropping this function's result is preffered in Zephyr I will change the code. Nevertheless I would vote for changing the API then :) 

Regarding the tracing, if `CONFIG_TRACING` is disabled the `SYS_PORT_TRACING_OBJ_FUNC` macro is defined as:

```
#define SYS_PORT_TRACING_OBJ_FUNC(obj_type, func, obj, ...) do { } while (false)
```

So it should not cause issues. Am I missing some puzzle here?

bbilas replied:
> Wouldn't it be better to ignore the return value of `k_thread_name_set`? That way it will work fine when thread naming is off but tracing is on; and when both are on or off.

For some reason, the `k_thread_name_set` function returns a specific return code we should handle. The suggested solution with `IS_ENABLED` macro is pretty enough for me.

GeorgeCGV replied:
> Nevertheless I would vote for changing the API then :)

+1

bbilas replied:
I've looked at the other PRs and Zephyr sources and it seems that the more common approach is to simply ignore the return code from `k_thread_name_set` even if thread names are enabled. This is a bit of nonsense and I agree that changes need to be made to the API.

bbilas commented:
cc @jeppenodgaard

LukaszMadejGrinn commented:
> I think #68048 should get some attention before this PR (I might be a bit subjective since I made that PR). These improvements will conflict with that PR because IC specific functions are moved into another function. I am looking for someone with access to a AD5592 to test that PR. Are you able to do that @LukaszMadejGrinn ?

I will try to test and share results tomorrow. Stay tuned.

jeppenodgaard commented:
@LukaszMadejGrinn https://github.com/zephyrproject-rtos/zephyr/pull/68048 has been merged so this PR needs to be rebased.

LukaszMadejGrinn committed:
drivers: adc: adc_ad559x: add invernal vref value

Add configuration of internal refference voltage value so raw ADC
readings can be converted to mV utilizing adc_raw_to_millivolts()
ADC API function.

Signed-off-by: Lukasz Madej <l.madej@grinn-global.com>

LukaszMadejGrinn committed:
drivers: adc: adc_ad559x: fix driver init

k_thread_name_set() function returns an error if CONFIG_THREAD_NAME is
not enabled which may make the driver not functional.

Fix the issue by adding a check against CONFIG_THREAD_NAME avaliability
before calling k_thread_name_set() function so the driver can operate
properly with CONFIG_THREAD_NAME disabled.

Signed-off-by: Lukasz Madej <l.madej@grinn-global.com>

LukaszMadejGrinn committed:
drivers: adc: adc_ad559x: improve adc read

Use information encoded in conversion result value to validate it.
Check if MSB bit is set to zero.
Check if channel number included in the result matches channel
number selected for conversion.
Use bitmask to extract converted value instead of math calculations.

Signed-off-by: Lukasz Madej <l.madej@grinn-global.com>

LukaszMadejGrinn commented:
Hi @anangl,

Can you please accept the PR so it can be merged?
